include/ignition/gazebo/detail/EntityComponentManager.hh

```diff
@@ -20,15 +20,61 @@
 #include <cstring>
 #include <map>
 #include <set>
+#include <type_traits>
 #include <utility>
 #include <vector>
 
+#include <ignition/math/Helpers.hh>
+
 #include "ignition/gazebo/EntityComponentManager.hh"
 
 namespace ignition
 {
 namespace gazebo
 {
+//////////////////////////////////////////////////
+namespace traits
+{
+  /// \brief Helper struct to determine if an equality operator is present.
+  struct TestEqualityOperator
+  {
+  };
+  template<typename T>
+  TestEqualityOperator operator == (const T&, const T&);
+
+  /// \brief Type trait that determines if an operator== is defined for `T`.
+  template<typename T>
+  struct HasEqualityOperator
+  {
+    enum
+    {
+      value = !std::is_same<decltype(*(T*)(0) == *(T*)(0)),
+          TestEqualityOperator>::value
+    };
+  };
+}
+
+//////////////////////////////////////////////////
+/// \brief Helper function to compare two objects of the same type using its
+/// equality operator.
+/// If `DataType` doesn't have an equality operator defined, it will return
+/// true.
+/// For doubles, `ignition::math::equal` will be used.
+template<typename DataType>
+auto CompareData = [&](const DataType &_a, const DataType &_b) -> bool
+{
+  if constexpr (std::is_same<DataType, double>::value)
+  {
+    return math::equal(_a, _b);
+  }
+  else if constexpr (traits::HasEqualityOperator<DataType>::value)
+  {
+    return _a == _b;
+  }
+
+  return false;
```

Comment:
Doc-comment says it returns true...

Reply:
oops, fixed: cb3e63a229666f88dd433afec03c6f6a49856b02